src/main.rs

```diff
@@ -1,28 +1,177 @@
+    forward_proxy::DEFAULT_PORT
+);
+
+const REVERSE_PORT_HELP : &str = const_format::formatcp!(
+    "port number receiving incoming connections, default {}",
+    reverse_proxy::DEFAULT_PORT
+);
+
+fn run() -> Result<()> {
+    let m = App::new(APP_NAME)
+        .about(ABOUT_STR)
+        .setting(AppSettings::SubcommandRequiredElseHelp)
+        .arg(Arg::with_name("compress")
+             .short("c")
+             .long("compress")
+             .help("enable compression"))
+        .arg(Arg::with_name("encrypt")
+             .short("e")
+             .long("encrypt")
+             .help("enable encryption"))
+        .subcommands( vec![ 
+            SubCommand::with_name("forward")
+                .about("start in foward proxy server mode")
+                .arg(Arg::with_name("port")
+                     .short("p")
+                     .long("port")
+                     .help(FORWARD_PORT_HELP)
+                     .takes_value(true)),
+
+            SubCommand::with_name("reverse")
+                .about("start in reverse proxy server mode")
+                .arg(Arg::with_name("port")
+                     .short("p")
+                     .long("port")
+                     .help(REVERSE_PORT_HELP)
+                     .takes_value(true))
+                .arg(Arg::with_name("SERVERS")
+                     .help("server addresses in format ip:port")
+                     .required(true)
+                     .multiple(true))
+        ])
+        .get_matches_safe()
+        .chain_err(|| "error parsing arguments")?;
+
+    let compress = m.is_present("compress");
+
+    let encrypt = m.is_present("encrypt");
+    
+    let server = match m.subcommand() {
+        ("forward", Some(sub_m)) => Server::Forward {
+            port: match sub_m.value_of("port") {
+                Some(p) => p.parse()
+                    .chain_err(
+                        || format!("error parsing port number \"{}\"", p)
+                    )?,
+                None => forward_proxy::DEFAULT_PORT,
+            },
+        },
+        ("reverse", Some(sub_m)) => Server::Reverse {
+            port: match sub_m.value_of("port") {
+                Some(p) => p.parse()
+                    .chain_err(
+                        || format!("error parsing port number \"{}\"", p)
+                    )?,
+                None => reverse_proxy::DEFAULT_PORT,
+            },
+            server_ips: match sub_m.values_of("SERVERS") {
+                Some(addrs) => addrs.map(
+                    |a| a.parse::<SocketAddr>()
+                        .chain_err(
+                            || format!("error parsing socket address \"{}\"", a)
+                        )
+                    ) 
+                    .collect::<Result<Vec<_>>>()?,
+                None => bail!("no server addreses"),
+            },
+        },
+        _ => bail!("unknown subcommand"),
+    };
+
+    let local_ip = "127.0.0.1"
```

Comment:
Here you can just use `IpV4Addr::new(127, 0, 0, 1)` since we're just parsing a constant (and this should never fail).